src/main/java/com/artipie/asto/bench/BenchmarkStorage.java

```diff
@@ -0,0 +1,144 @@
+import java.util.Collections;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionStage;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentSkipListMap;
+import java.util.function.Function;
+
+/**
+ * Storage implementation for benchmarks.
+ * @since 1.1.0
+ * @checkstyle ClassDataAbstractionCouplingCheck (500 lines)
+ */
+@SuppressWarnings("PMD.AvoidDuplicateLiterals")
+public final class BenchmarkStorage implements Storage {
+    /**
+     * Backend storage.
+     */
+    private final InMemoryStorage backend;
+
+    /**
+     * Local storage.
+     */
+    private final Map<Key, byte[]> local;
+
+    /**
+     * Set which contains deleted keys.
+     */
+    private final Set<Key> deleted;
+
+    /**
+     * Ctor.
+     * @param backend Backend storage
+     */
+    public BenchmarkStorage(final InMemoryStorage backend) {
+        this(backend, Collections.emptyMap());
+    }
+
+    /**
+     * Ctor.
+     * @param backend Backend storage
+     * @param local Content of local storage
+     */
+    public BenchmarkStorage(final InMemoryStorage backend, final Map<Key, byte[]> local) {
+        this.backend = backend;
+        this.local = new ConcurrentSkipListMap<>(local);
+        this.deleted = ConcurrentHashMap.newKeySet();
+    }
+
+    @Override
+    public CompletableFuture<Boolean> exists(final Key key) {
+        throw new NotImplementedError("Not implemented yet");
+    }
+
+    @Override
+    public CompletableFuture<Collection<Key>> list(final Key prefix) {
+        throw new NotImplementedError("Not implemented yet");
+    }
+
+    @Override
+    public CompletableFuture<Void> save(final Key key, final Content content) {
+        final CompletableFuture<Void> res;
+        if (Key.ROOT.equals(key)) {
+            res = new CompletableFutureSupport.Failed<Void>(
+                new ArtipieIOException("Unable to save to root")
+            ).get();
+        } else {
+            res = new Concatenation(new OneTimePublisher<>(content)).single()
+                .to(SingleInterop.get())
+                .thenApply(Remaining::new)
+                .thenApply(Remaining::bytes)
+                .thenAccept(bytes -> this.local.put(key, bytes))
+                .toCompletableFuture();
+        }
+        return res;
+    }
+
+    @Override
+    public CompletableFuture<Void> move(final Key source, final Key destination) {
+        throw new NotImplementedError("Not implemented yet");
+    }
+
+    @Override
+    public CompletableFuture<Long> size(final Key key) {
+        throw new NotImplementedError("Not implemented yet");
+    }
+
+    @Override
+    public CompletableFuture<Content> value(final Key key) {
+        return CompletableFuture.runAsync(
+            () -> {
+                if (this.deleted.contains(key)) {
+                    throw new ValueNotFoundException(key);
+                }
```

Comment:
We can perform this check synchronously without submitting a new task in pool, set check for key is really fast. It may decrease benchmark's baseline.